DirectProgramming/DPC++FPGA/PlatformTests/board_test/src/board_test.hpp

```diff
@@ -0,0 +1,1852 @@
+// is working fine
+
+size_t ShimMetrics::TestGlobalMem(queue &q) {
+  // Data is transferred from host to device in kMaxHostChunk size transfers
+  // (size in bytes)
+  constexpr size_t kMaxHostChunk = 512 * kMB;
+
+  // Test fails if max alloc size is 0
+  if (max_alloc_size == 0) {
+    std::cout << "Maximum global memory allocation supported by Sycl device is "
+              << "0! Cannot run kernel-to-memory read wite test\n\n";
+    return 1;
+  }
+
+  // **** Create device buffer ****//
+  // Creating device buffer to span all usable global memory space on device
+  buffer<unsigned long, 1> dev_buf{
+      range<1>{(max_alloc_size / sizeof(unsigned long))}};
+  std::cout << "Size of buffer created = " << dev_buf.get_size() << " bytes\n";
+
+  // **** Host memory allocation **** //
+  /*<hostbuf> is used to store data
+  to be written to device buffer as well as data that is read back
+  While loop retries allocation for smaller chunk if kMaxHostChunk allocation
+  fails */
+  // Size of allocated host memory (in bytes)
+  size_t host_size = kMaxHostChunk;
+  unsigned long *hostbuf = (unsigned long *)malloc(host_size);
+  while (hostbuf == NULL) {
+    host_size = host_size / 2;
+    hostbuf = (unsigned long *)malloc(host_size);
+  }
+
+  // **** Writing data from host memory to device global memory **** //
+
+  // Number of bytes remaining to transfer to device memory
+  unsigned long bytes_rem = max_alloc_size;
+  // offset at which write should begin in device global memory
+  unsigned long offset = 0;
+  // Number of chunks of host_size written to device
+  unsigned long chunk_cnt_wr = 0;
+  // Total time to write
+  double sum_time_ns = 0.0;
+
+  // Copying host memory to device buffer in chunks
+  std::cout << "Writing " << (max_alloc_size / kMB)
+            << " MB to device global memory ... ";
+
+  // Device global memory larger (i.e. max_alloc_size) than host memory size
+  // (i.e. host_size) Chunks of host memory size written to device memory in
+  // iterations
+  while (bytes_rem > 0) {
+    unsigned long chunk = bytes_rem;
+    if (chunk > host_size) chunk = host_size;
+
+    // Initializing host buffer
+    for (unsigned long i = 0; i < chunk / sizeof(unsigned long); ++i) {
+      hostbuf[i] = offset + i;
+    }
+
+    // Submit command to copy (explicit copy from host to device)
+    auto h2d_copy_e = q.submit([&](handler &h) {
+      // Range of buffer that needs to accessed
+      auto buf_range = chunk / sizeof(unsigned long);
+      // offset starts at 0 - incremented by chunk size each iteration
+      auto buf_offset = offset / sizeof(unsigned long);
+      // Access host_size range of buffer starting at buf_offset
+      accessor mem(dev_buf, h, buf_range, buf_offset);
+      // Writing from host memory to device buffer
+      h.copy(hostbuf, mem);
+    });
+
+    // Wait for explicit copy from host memory to device buffer to complete
+    h2d_copy_e.wait();
```

Comment:
you can put the `.wait()` on the submit. Save some code :)

Reply:
Cannot use this for submit calls that need to return an event used to measure kernel execution time. Have made the change to submit calls that do not need to return event type.